timemachine/cpp/src/bd_exchange_move.cu

```diff
@@ -0,0 +1,186 @@
+    gpuErrchk(cudaMemset(d_log_sum_exp_before_.data, 0, d_log_sum_exp_before_.size()));
+    gpuErrchk(cudaMemset(d_log_sum_exp_after_.data, 0, d_log_sum_exp_after_.size()));
+    curandErrchk(curandCreateGenerator(&cr_rng_, CURAND_RNG_PSEUDO_DEFAULT));
+    curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+}
+
+template <typename RealType> BDExchangeMove<RealType>::~BDExchangeMove() {
+    curandErrchk(curandDestroyGenerator(cr_rng_));
+}
+
+template <typename RealType>
+void BDExchangeMove<RealType>::move_device(
+    const int N,
+    const int num_moves,
+    double *d_coords, // [N, 3]
+    double *d_box,    // [3, 3]
+    cudaStream_t stream) {
+
+    if (N != N_) {
+        throw std::runtime_error("N != N_");
+    }
+
+    // Set the stream for the generator
+    curandErrchk(curandSetStream(cr_rng_, stream));
+
+    const int tpb = DEFAULT_THREADS_PER_BLOCK;
+    const int mol_blocks = ceil_divide(num_target_mols_, tpb);
+    // Compute logsumexp of energies once upfront to get log probabilities
+    mol_potential_.mol_energies_device(
+        N,
+        num_target_mols_,
+        d_coords,
+        d_params_.data,
+        d_box,
+        d_mol_energy_buffer_.data, // Don't need to zero, will be overridden
+        stream);
+
+    // Don't need to normalize to sample
+    k_compute_log_weights_from_energies<RealType><<<mol_blocks, tpb, 0, stream>>>(
+        num_target_mols_, beta_, d_mol_energy_buffer_.data, d_log_weights_before_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    logsumexp_.sum_device(num_target_mols_, d_log_weights_before_.data, d_log_sum_exp_before_.data, stream);
+
+    const int num_samples = 1;
+    for (int move = 0; move < num_moves; move++) {
+        // Run only after the first pass, to maintain meaningful `log_probability_host` values
+        if (move > 0) {
+            // Run a separate kernel to replace the before log probs and weights with the after if accepted a move
+            // Need the weights to sample a value and the log probs are just because they aren't expensive to copy
+            k_store_accepted_log_probability<RealType><<<1, tpb, 0>>>(
+                num_target_mols_,
+                d_translations_.data + 3, // Offset to get the last value for the acceptance criteria
+                d_log_sum_exp_before_.data,
+                d_log_sum_exp_after_.data,
+                d_log_weights_before_.data,
+                d_log_weights_after_.data);
+            gpuErrchk(cudaPeekAtLastError());
+        }
+        // Make a copy of the coordinates
+        gpuErrchk(cudaMemcpyAsync(
+            d_intermediate_coords_.data, d_coords, d_intermediate_coords_.size(), cudaMemcpyDeviceToDevice, stream));
+
+        // Quaternions generated from normal noise, while translations and the acceptance value are uniform
+        curandErrchk(templateCurandNormal(cr_rng_, d_quaternions_.data, d_quaternions_.length, 0.0, 1.0));
```

Comment:
No batching yet, but so far this is like 0.5% of the total time. Its computing the per mol energies that are expensive.

Reply:
Yeah I don't expect this to be expensive relative to other bits of the code.